fl4health/preprocessing/autoencoders/loss.py

```diff
@@ -0,0 +1,73 @@
+from typing import Tuple
+
+import torch
+from torch.nn.modules.loss import _Loss
+
+
+class VAE_loss(_Loss):
+    """Implements the loss function used for training CVAEs and VAEs.
+    This loss computes the base_loss (defined by the user) between the input and generated output.
+    It then adds the KL divergence between the estimated distribution (represented by mu and logvar)
+    and the standard normal distribution.
+    """
+
+    def __init__(
+        self,
+        latent_dim: int,
+        base_loss: _Loss = torch.nn.BCELoss(reduction="sum"),
+    ) -> None:
+        """Initializes the VAE loss function.
+
+        Args:
+            latent_dim (int): Dimensionality of the latent space.
+            base_loss (_Loss, optional): Base loss function. Defaults to torch.nn.BCELoss(reduction="sum").
+        """
+        super().__init__()
+        # User can define a base_loss to measure the distance between the input and generated output.
+        self.base_loss = base_loss
+        # Latent dimension is used to unpack the model output
+        self.latent_dim = latent_dim
+
+    def kl_divergence_loss(self, mu: torch.Tensor, logvar: torch.Tensor) -> torch.Tensor:
+        """Calculates the KL divergence loss.
+
+        Args:
+            mu (torch.Tensor): Mean of the estimated distribution.
+            logvar (torch.Tensor): Log variance of the estimated distribution.
+
+        Returns:
+            torch.Tensor: KL divergence loss.
+        """
+        kl_divergence_loss = -0.5 * torch.sum(1 + logvar - mu.pow(2) - logvar.exp())
+        return kl_divergence_loss
+
+    def unpack_model_output(self, preds: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
+        """Unpacks the model output tensor.
+
+        Args:
+            preds (torch.Tensor): Model predictions.
+
+        Returns:
+            Tuple[torch.Tensor, torch.Tensor, torch.Tensor]: Unpacked output containing predictions, mu, and logvar.
+        """
+        # The order of logvar and mu in the output tensor is important.
+        # For each model output, the first self.latent_dim indices are used to store the log variance,
+        # the next self.latent_dim indices are allocated to mu, and the remaining indices store the model predictions.
+        logvar = preds[:, 0 : self.latent_dim]
+        mu = preds[:, self.latent_dim : 2 * self.latent_dim]
+        preds = preds[:, 2 * self.latent_dim :]
```

Comment:
There are a few assumptions (correct me if I'm wrong on the assumptions below) that are being made here that we want to surface in the comments for clarity, I think. The first is that the preds are "batch first". The second is that they are "flattened." That is, the preds tensors are of dimension 2. We might want to assert that the dimension of preds is two and log an informative error if it is not.